checker.go

```diff
@@ -527,6 +528,75 @@ func (c *allChecker) Check(got interface{}, args []interface{}, notef func(key s
 	return nil
 }
 
+// JSONEquals is a checker that checks whether a byte slice
+// or string is JSON-equivalent to a Go value. See CodecEquals for
+// more information.
+//
+// It uses DeepEquals to do the comparison. If a more sophisticated
+// comparison is required, use CodecEquals directly.
+var JSONEquals = CodecEquals(json.Marshal, json.Unmarshal)
+
+type codecEqualChecker struct {
+	marshal    func(interface{}) ([]byte, error)
+	unmarshal  func([]byte, interface{}) error
+	deepEquals Checker
+}
+
+// CodecEquals returns a checker that checks for codec value equivalence.
+//
+// It expects two arguments: a byte slice or a string containing some codec-marshaled
+// data, and a Go value.
+//
+// It uses unmarshal to unmarshal the data into an interface{} value.
+// It marshals the Go value using marshal, then unmarshals the result into
+// an interface{} value.
+//
+// It then checks that the two interface{} values are deep-equal to one another,
+// using CmpEquals(opts) to perform the check.
+//
+// See JSONEquals for an example of this in use.
+func CodecEquals(
+	marshal func(interface{}) ([]byte, error),
+	unmarshal func([]byte, interface{}) error,
+	opts ...cmp.Option,
+) Checker {
+	return &codecEqualChecker{
+		marshal:    marshal,
+		unmarshal:  unmarshal,
+		deepEquals: CmpEquals(opts...),
+	}
+}
+
+func (c *codecEqualChecker) ArgNames() []string {
+	return []string{"got", "want"}
+}
+
+func (c *codecEqualChecker) Check(got interface{}, args []interface{}, note func(key string, value interface{})) error {
+	var gotContent []byte
+	switch got := got.(type) {
+	case string:
+		gotContent = []byte(got)
+	case []byte:
+		gotContent = got
+	default:
+		return BadCheckf("expected string or byte, got %T", args[0])
+	}
+	expectContent := args[0]
+	expectContentBytes, err := c.marshal(expectContent)
+	if err != nil {
+		return BadCheckf("cannot marshal expected contents: %v", err)
+	}
+	var expectContentVal interface{}
+	if err := c.unmarshal(expectContentBytes, &expectContentVal); err != nil {
+		return BadCheckf("cannot unmarshal expected contents: %v", err)
+	}
+	var gotContentVal interface{}
+	if err := c.unmarshal([]byte(gotContent), &gotContentVal); err != nil {
+		return fmt.Errorf("cannot unmarshal obtained contents: %v; %q", err, gotContent)
```

Comment:
SHould this be a BadCheck as well? If you cannot unmarshal a value, that value cannot either equal or not equal the want value.

Reply:
I don't think so - a BadCheck is about the expected value being bad, not the data itself.